verification/spec/connection-handshake/Environment.tla

```diff
@@ -0,0 +1,402 @@
+            arbitraryMsg \in {msg \in ConnectionHandshakeMessages : 
+                msg.type = "ICS3MsgTry"}}
+\*            arbitraryMsg \in ConnectionHandshakeMessages
+\*            /\ arbitraryMsg.type = "ICS3MsgTry"}
+       /\ UNCHANGED inBufChainA
+
+
+(* Environment next action.
+
+TODO: explain the relaying functionality and additional variables.
+
+    There are four possible actions that the environment may perform:
+
+    1. A good step: the environment advances or updates the client on
+    one of the two chains.
+
+    2. The environment becomes malicious, as a result of both chains
+    advancing past the UNINIT step (i.e., after both chains finished 
+    locking on to a set of connection parameters).
+
+    3. A malicious step.
+
+    4. The environment stops acting maliciously.
+ *)
+NextEnv ==
+    \/ /\ GoodNextEnv                               (* A good step. *)
+       /\ UNCHANGED maliciousEnv
+    \/ /\ RelayNextEnv
+       /\ UNCHANGED maliciousEnv
+\*    \/ /\ UNCHANGED <<chainAVars, chainBVars, maliciousEnv>>
+\*    \/ /\ ~ maliciousEnv                            (* Enable malicious env. *)
+\*       /\ storeChainA.connection.state # "UNINIT"
+\*       /\ storeChainB.connection.state # "UNINIT"
+\*       /\ maliciousEnv' = TRUE
+\*       /\ MaliciousNextEnv
+\*       /\ UNCHANGED<<chainStoreVars, outBufChainA, outBufChainB>>
+\*    \/ /\ maliciousEnv                              (* A malicious step. *)
+\*       /\ MaliciousNextEnv
+\*       /\ UNCHANGED<<maliciousEnv, chainStoreVars, outBufChainA, outBufChainB>>
+\*    \/ /\ maliciousEnv                              (* Disable malicious env. *)
+\*       /\ maliciousEnv' = FALSE
+\*       /\ UNCHANGED<<chainAVars, chainBVars>>
+
+
+(* Enables when the connection is open on both chains.
+
+    State predicate signaling that the protocol terminated correctly.
+ *)
+ICS3ReachTermination ==
+    /\ storeChainA.connection.state = "OPEN"
+    /\ storeChainB.connection.state = "OPEN"
+    /\ UNCHANGED allVars
+
+
+
+ICS3NonTermination ==
+    /\ \/ (~ chmA!CanAdvance /\ storeChainA.connection.state # "OPEN")
+       \/ (~ chmB!CanAdvance /\ storeChainB.connection.state # "OPEN")
+    /\ UNCHANGED allVars
+
+(******************************************************************************
+
+    Main spec. The system comprises the environment plus the two instances of
+    ICS3 modules.
+
+ *****************************************************************************)
+
+
+(* Initializes both chains, attributing to each a chainID and a client. *)
+Init ==
+    /\ \E clientA \in InitClients({ x.clientID : x \in ChainAConnectionEnds }) :
+            chmA!Init("chainA", clientA, NullConnection)
+    /\ \E clientB \in InitClients({ x.clientID : x \in ChainBConnectionEnds }) :
+            chmB!Init("chainB", clientB, NullConnection)
+    /\ InitEnv
+
+
+(* The two ICS3 modules and the environment alternate their steps. *)
+Next ==
+    \/ ICS3ReachTermination
```

Comment:
I think we normally want this logic to be part of precondition for something to happen: for example if it's not true then we make a step; otherwise we don't do anything. I think that it is semantically slightly different to what we have here.